### PR TITLE
Add authorization check to get process

### PR DESCRIPTION
**Pull request by davewalter**
## Is there a related GitHub Issue?
#554

## What is this change about?
This PR adds an authorization check to the GET /v3/processes/:guid endpoint.

## Does this PR introduce a breaking change?
No

## Acceptance Steps
See #554 

## Tag your pair, your PM, and/or team
@Birdrock

### CODE REVIEW

File: api/repositories/process_repository.go [hunk 1]
```diff
@@ -102,7 +110,27 @@ func (r *ProcessRepo) GetProcess(ctx context.Context, authInfo authorization.Inf
 	if err != nil { // untested
 		return ProcessRecord{}, err
 	}
-	return returnProcess(processList.Items)
```
Comment (danail-branekov) [inline api/repositories/process_repository.go:105]: I would suggest renaming `returnProcess` to `ensureSingleProcess` or something similar. `returnProcess` sounds like a `return` statement to me and I would expect no code after it.

File: api/repositories/process_repository_test.go [hunk 2]
```diff
@@ -4,131 +4,142 @@ import (
...
 
-		app1GUID = generateGUID()
-		app2GUID = generateGUID()
-		process1GUID = generateGUID()
-		process2GUID = generateGUID()
+		spaceDeveloperClusterRole = createClusterRole(testCtx, SpaceDeveloperClusterRoleRules)
 	})
 
 	AfterEach(func() {
 		Expect(
```
Comment (danail-branekov) [inline api/repositories/process_repository_test.go:62]: I believe that for env tests cleaning up the namespaces makes no sense, as the test env gets destroyed anyway. Also, as in env tests there are no controllers, deleting the namespace does not delete the objects it contains, they remain as records in the etcd.
TL;DR I don't think this `AfterEach` is needed at all.

File: api/repositories/process_repository_test.go [hunk 3]
```diff
@@ -4,131 +4,142 @@ import (
...
+			cfApp1 = initializeAppCR("test-app1", app1GUID, space1.Name)
+			Expect(k8sClient.Create(context.Background(), cfApp1)).To(Succeed())
+
+			cfApp2 = initializeAppCR("test-app2", app2GUID, space2.Name)
+			Expect(k8sClient.Create(context.Background(), cfApp2)).To(Succeed())
 
-			namespace2Name := generateGUID()
-			namespace2 = &corev1.Namespace{ObjectMeta: metav1.ObjectMeta{Name: namespace2Name}}
-			Expect(k8sClient.Create(context.Background(), namespace2)).To(Succeed())
+			cfProcess1 = initializeProcessCR(process1GUID, space1.Name, app1GUID)
```
Comment (danail-branekov) [inline api/repositories/process_repository_test.go:88]: Are there usages of the initializeXXXCR functions that are not followed by `k8sClient.Create`? If not, I would suggest putting CR initialisation and its creation into a single function and internally ensures that cration succeeds - that would spare all the creates in tests.

File: api/tests/e2e/processes_test.go [hunk 4]
```diff
@@ -0,0 +1,74 @@
...
+		appGUID     string
+		processGUID string
+		result      resource
+		list        resourceList
+		resp        *resty.Response
+	)
+
+	BeforeEach(func() {
+		orgGUID = createOrg(generateGUID("org"))
+		DeferCleanup(func() {
```
Comment (danail-branekov) [inline api/tests/e2e/processes_test.go:25]: I personally find `DeferCleanup` a bit hard to follow as I would not expect cleanup code in `BeforeEach`. I would suggest moving org deletion into a `AfterEach`

File: api/repositories/process_repository_test.go [hunk 5]
```diff
@@ -4,131 +4,142 @@ import (
...
+			Expect(k8sClient.Create(context.Background(), cfProcess2)).To(Succeed())
 		})
 
-		AfterEach(func() {
-			Expect(k8sClient.Delete(context.Background(), namespace2)).To(Succeed())
+		JustBeforeEach(func() {
+			process, getErr = processRepo.GetProcess(testCtx, authInfo, process1GUID)
 		})
 
 		When("on the happy path", func() {
```
Comment (danail-branekov) [inline api/repositories/process_repository_test.go:99]: Now that we have authrorized vs unauthorized case, I would get rid of the `happy path` context and put the `unauthorized` it in the "default" context (as by default one should get authorization error). And then move the "happy path" under `When("the user is a SpaceDeveloper", ....`

File: api/repositories/process_repository_test.go [hunk 6]
```diff
@@ -138,87 +149,56 @@ var _ = Describe("ProcessRepo", func() {
...
+			Expect(k8sClient.Create(context.Background(), cfProcess1)).To(Succeed())
 
-			cfProcess2 = initializeProcessCR(process2GUID, namespace.Name, app1GUID)
-			Expect(k8sClient.Create(testCtx, cfProcess2)).To(Succeed())
+			cfProcess2 = initializeProcessCR(process2GUID, space1.Name, app1GUID)
+			Expect(k8sClient.Create(context.Background(), cfProcess2)).To(Succeed())
 		})
 
 		When("on the happy path", func() {
 			When("spaceGUID is not empty", func() {
```
Comment (danail-branekov) [inline api/repositories/process_repository_test.go:193]: I believe that this is the default case, it should no be inside a dedicated context

File: api/repositories/process_repository_test.go [hunk 7]
```diff
@@ -138,87 +149,56 @@ var _ = Describe("ProcessRepo", func() {
...
 				It("returns Process records for the AppGUID we request", func() {
-					processes, err := processRepo.ListProcesses(testCtx, authInfo, ListProcessesMessage{AppGUID: []string{app1GUID}, SpaceGUID: namespace.Name})
+					processes, err := processRepo.ListProcesses(testCtx, authInfo, ListProcessesMessage{AppGUID: []string{app1GUID}, SpaceGUID: space1.Name})
```
Comment (danail-branekov) [inline api/repositories/process_repository_test.go:195]: I believe it would more Ginkgo-idiomatic to put the function under test (i.e. `List`) in `JustBeforeEach` instead of calling it in every `It`

**Comment by danail-branekov** [comment]
It seems that PR https://github.com/cloudfoundry/cf-k8s-controllers/pull/637 is conflicting a lot with this one. I am afraid that there would be lots of merge conflicts :(

**Comment by danail-branekov** [comment]
Hey @davewalter 
I have rebased the pr on top of main for you. It is unfair to make you do that on a cross-team collaboration story. 
Hopefully next time we pick a cross-team story for you we would be careful not to heavily modify the code the story is all about so that we do not step on each others' toes.

I think that most of my comments have been already addressed with PR #637 , with the exception of the discussion on dot importing non-ginkgo packages.

File: api/repositories/process_repository_test.go (hunk 3 above)
Comment (davewalter) [reply api/repositories/process_repository_test.go]: We've created a separate chore to address this across the entire set of repository tests: #647

**Comment by julian-hj** [comment]
Thanks @kieron-dev! I see that @gcapizzi has already re-opened the github issue. We'll take another crack at it in a separate PR.
